src/grammar.rs

```diff
@@ -155,6 +207,44 @@ enum Expr {
     Predefined(Predefined),
 }
 
+impl Expr {
+    /// See [`Grammar::how_many`].
+    ///
+    /// `mem` is previously calculated `how_many` for `depth - 1` for each rule in `rules`.
+    fn how_many(&self, rules: &[Expr], depth: usize, mem: &[Option<u64>]) -> Option<u64> {
```

Comment:
understanding: What does the `depth` parameter do in this context? Is it just passed to recursive calls?